expression/parser.py

```diff
@@ -242,14 +242,26 @@ def visit_Expr(self, node):
     def visit_BoolOp(self, node):
         """
         Visit a boolean expression node.
+
+        Implement proper short-circuit semantics instead of eagerly
+        evaluating all operands.
         """
 
         op = type(node.op)
-        func = self._boolean_ops[op]
-        result = func(self.visit(node.values[0]), self.visit(node.values[1]))
-        for value in node.values[2:]:
-            result = func(result, self.visit(value))
-
+        # Short-circuit for 'and'
+        if op is ast.And:
+            result = self.visit(node.values[0])
+            for value in node.values[1:]:
+                if not result:
+                    return result
+                result = self.visit(value)
+            return result
+        # Short-circuit for 'or'
+        result = self.visit(node.values[0])
+        for value in node.values[1:]:
+            if result:
+                return result
+            result = self.visit(value)
         return result
```

Comment:
```suggestion
        # Short-circuit for 'and'
        if op is ast.And:
            result = False
            for value in node.values:
                result = self.visit(value)
                if not result:
                    return result
            return result
        # Short-circuit for 'or'
        result = True
        for value in node.values:
            result = self.visit(value)
            if result:
                return result
        return result
```
I feel like this flows better and reduces the number of `visit` calls even more. It does need a dummy default for the result to not have linters think the variable may be undefined, but that also makes the difference between 'and' and 'or' more clear.

Reply:
done